includes/class-llms-assets.php

```diff
@@ -0,0 +1,343 @@
+<?php
+/**
+ * Methods for static asset registration and enqueueing
+ *
+ * These methods require assets to be "defined" in a structured format.
+ *
+ * A defined asset is then enqueued or registered with the WordPress core using this derivative
+ * API that requires only script handles.
+ *
+ * This API also aims to reduce redundancy in asset registrations by allowing "partial" definitions
+ * which are filled with default values. For example, every asset in LifterLMS shares the same base
+ * plugin url. Using this API we define that URL one time, instead of defining it over and over for
+ * each individual asset.
+ *
+ * @package LifterLMS/Classes
+ *
+ * @since [version]
+ * @version [version]
+ */
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * LLMS_Assets Class
+ *
+ * @since [version]
+ */
+class LLMS_Assets {
```

Comment:
do we plan to extend this?
If not, let's prevent it, no?
```suggestion
final class LLMS_Assets {
```

and of course `protected` methods do not make much sense in that case => `private`.

If, dunno, you're planning to extend this class in our add-ons then the `$scripts` and `$styles` properties, as well as the methods that manipulate them, shouldn't be static, right?
In this case `defaults` would be values passed during the instantiation of this class in class-lifterlms.php and assign the instance to a LifterLMS property, like `$session` and `$query`. What do you think?

Just considerations that, obviously you can totally skip <3

Reply:
I do want this class to be extendable so that add-ons by us or 3rd parties can easily define their own assets and use these methods; should they wish.

My thinking was that the method by which add-ons would utilize this class would not be to extend the `$scripts` and `$styles` properties, but to instead utilize any number of the filters run by this class to define their own set of assets.

Though now that you mention this, maybe it would make sense to allow definitions to be added to these arrays.

Converting to a singleton as you're suggesting seems like a good way to do this. In that case I think everything should be converted from static (all the methods) as they'd be accessible via the singleton as you've suggested.

A developer could then extend the the class (I don't want to make it final with private methods) or they could use the filters. Depending on the complexity of the project or the number of assets one may make sense over the other.

An add-on with a single asset like a CSS file may just want to define it with a filter whereas a more complicated add-on may wish to extend the class, create their own asset list, and add them that way.

I think in this scenario we should probably create a `define()` method which would accept an array that would be added to the the relevant asset list.

Reply:
I actually wasn't suggesting to make it a singleton 👍 
I only wanted the LifterLMS class to instantiate its own and assign it to a property that can be referenced from other parts of the code.

I don't exactly remember the details but I remember that extending a singleton was not a good practice...

Reply:
Oh I see what you're saying. That would mean fewer rewritten tests. I'll update the PR in a bit